Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean

```diff
@@ -711,6 +716,67 @@ lemma toSpec_bijective {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m):
 
 end fromSpecToSpec
 
+variable {𝒜} in
+/--The continuous function `Spec A⁰_f → Proj|D(f)` by sending `q` to `{a | aᵢᵐ/fⁱ ∈ q}`.-/
+def fromSpec {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m) :
+    (Spec.T (A⁰_ f)) ⟶ (Proj.T| (pbo f)) where
+  toFun := FromSpec.toFun f_deg hm
+  continuous_toFun :=
+    (IsTopologicalBasis.continuous_iff <|
```

Comment:
I am sorry for taking such a long time to review this. This is a great result. However, I feel that the readability could be improved. Some auxiliary lemmas which may be useful could be extracted from the proof also.
The following general topology lemma could be introduced (if mathlib experts in topology are ok with it):
```lean
lemma _root_.TopologicalSpace.IsTopologicalBasis.subtype
    {α : Type*} [TopologicalSpace α] {B : Set (Set α)}
    (h : TopologicalSpace.IsTopologicalBasis B) (p : α → Prop) :
    IsTopologicalBasis (Set.preimage (Subtype.val (p := p)) '' B) :=
  h.inducing (f := Subtype.val (p := p)) ⟨rfl⟩
```
The following may not be very useful as is, but could be generalized for `Equiv` (if we do not have it already!):
```lean
variable {𝒜} in
lemma preimage_fromSpec_eq_image_toSpec {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m) :
    Set.preimage (FromSpec.toFun f_deg hm) = Set.image toSpec := by
  sorry
```

Then, the skeleton of the proof could become
```lean
variable {𝒜} in
/--The continuous function `Spec A⁰_f → Proj|D(f)` by sending `q` to `{a | aᵢᵐ/fⁱ ∈ q}`.-/
def fromSpec {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m) :
    (Spec.T (A⁰_ f)) ⟶ (Proj.T| (pbo f)) where
  toFun := FromSpec.toFun f_deg hm
  continuous_toFun := by
    rw [((ProjectiveSpectrum.isTopologicalBasis_basic_opens 𝒜).subtype (pbo f).1).continuous_iff]
    rintro s ⟨_, ⟨a, rfl⟩, rfl⟩
    dsimp
    have h₁ : Subtype.val (p := (pbo f).1) ⁻¹' (pbo a) =
        ⋃ i : ℕ, {x | x.1 ∈ (pbo f) ⊓ pbo (decompose 𝒜 a i)} := sorry -- insert calc block here
    have h₂ : ∀ i, toSpec (f := f) '' {x | x.1 ∈ (pbo f) ⊓ pbo (decompose 𝒜 a i)} =
      (PrimeSpectrum.basicOpen (R := A⁰_ f) <|
        Quotient.mk'' ⟨m * i, ⟨decompose 𝒜 a i ^ m, SetLike.pow_mem_graded _ (Submodule.coe_mem _)⟩,
          ⟨f^i, by rw [mul_comm]; exact SetLike.pow_mem_graded _ f_deg⟩, ⟨i, rfl⟩⟩).1 := sorry
    rw [preimage_fromSpec_eq_image_toSpec]
    erw [h₁]
    rw [Set.image_iUnion]
    refine isOpen_iUnion fun i ↦ ?_
    erw [h₂]
    apply PrimeSpectrum.isOpen_basicOpen
```
Actually, the assertions `h₁` and `h₂` could be stated as separate lemmas.

Reply:
Thanks for the suggestions! I looked at other lemmas about `isTopologicalBasis` and arrived at conclusion that the lemma should be called `isTopologicalBasis_subtype`.

I have extracted `h_2` out as a separate lemma as it is useful (asserting that image of a basic open set is basic open). But I left `h_1` alone, for the lemma seems not that important.

You are right! `preimage_fromSpec_eq_image_toSpec` is in mathlib already.